lua/glance/list.lua

```diff
@@ -372,16 +394,22 @@ end
 
 function List:previous(opts)
   opts = opts or {}
-  for i = self:get_line() - 1 + (opts.skip or 0), 0, -1 do
-    local item = self.items[i]
-    if item and opts.loc_only and item.is_file then
-      if folds.is_folded(item.filename) then
-        self:toggle_fold(item)
-        return self:previous({ skip = item.count - 1 })
-      end
-      return self:previous({ skip = -1, loc_only = true })
+  for i, item in
+    self:walk({
+      start = self:get_line() + (opts.offset or 0),
+      cycle = opts.cycle,
+      backwards = true,
+    })
+  do
+    if opts.loc_only and item.is_file and folds.is_folded(item.filename) then
+      self:toggle_fold(item)
```

Comment:
when cycling back to a fold in the end of the list, cursor bounces in the wrong direction
we could do a check here before the toggle, if the previous line is the last item in the list
```lua
is_last_line = i == vim.api.nvim_buf_line_count(self.bufnr)
self:toggle_fold(item)
```

Reply:
Ah nice catch! I found another issue with the iterator where it would produce values that are out-of-bounds of the buffer. For example if you have collapsed all groups like so:

![image](https://user-images.githubusercontent.com/6705160/215447700-20ecd388-8813-4d77-8cee-7db5749816db.png)

and then navigate to the very bottom of the list and trigger the `Glance.actions.next` action (i.e. no cycling), it'll error by trying to put the cursor on the following, non-existent, line. Fixed via:

```diff
diff --git a/lua/glance/list.lua b/lua/glance/list.lua
index 9b15dd9..bd8d708 100644
--- a/lua/glance/list.lua
+++ b/lua/glance/list.lua
@@ -356,12 +356,13 @@ end
 function List:walk(opts)
   local idx = opts.start
   return function()
+    local line_count = vim.api.nvim_buf_line_count(self.bufnr)
     idx = idx + (opts.backwards and -1 or 1)
     if opts.cycle then
-      idx = ((idx - 1) % vim.api.nvim_buf_line_count(self.bufnr)) + 1
+      idx = ((idx - 1) % line_count) + 1
     end
     local item = self.items[idx]
-    if not item then
+    if not item or idx > line_count then
       return nil
     end
     return idx, item
```